contracts/stellar-interchain-token-service/src/flow_limit.rs

```diff
@@ -56,18 +56,24 @@ impl FlowDirection {
 
         ensure!(flow_amount <= flow_limit, ContractError::FlowLimitExceeded);
 
-        let new_flow = self
-            .flow(env, token_id.clone())
-            .checked_add(flow_amount)
+        let flow_amount_ = flow_amount as u128;
+        let flow_limit_ = flow_limit as u128;
+        let current_flow_ = self.flow(env, token_id.clone()) as u128;
+        let reverse_flow_ = self.reverse_flow(env, token_id.clone()) as u128;
+
+        let new_flow_ = current_flow_
+            .checked_add(flow_amount_)
             .ok_or(ContractError::FlowAmountOverflow)?;
-        let max_allowed = self
-            .reverse_flow(env, token_id.clone())
-            .checked_add(flow_limit)
+        let max_allowed_ = reverse_flow_
+            .checked_add(flow_limit_)
             .ok_or(ContractError::FlowAmountOverflow)?;
 
         // Equivalent to flow_amount + flow - reverse_flow <= flow_limit
-        ensure!(new_flow <= max_allowed, ContractError::FlowLimitExceeded);
+        ensure!(new_flow_ <= max_allowed_, ContractError::FlowLimitExceeded);
 
+        let new_flow: i128 = new_flow_
+            .try_into()
+            .map_err(|_| ContractError::FlowAmountOverflow)?;
```

Comment:
this can still error since `new_flow` could exceed `i128::MAX`, the changes just make the error happen on this line instead of earlier. So, the fundamental logic needs to be changed. And this needs some unit tests